File: app/src/test/java/com/farmerbb/taskbar/service/FavoriteAppTileServiceTest.java
package com.farmerbb.taskbar.service;

import android.app.Application;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Process;
import android.os.UserManager;
import android.service.quicksettings.Tile;

import androidx.test.core.app.ApplicationProvider;

import com.farmerbb.taskbar.R;
import com.farmerbb.taskbar.activity.PersistentShortcutLaunchActivity;
import com.farmerbb.taskbar.activity.PersistentShortcutSelectAppActivity;
import com.farmerbb.taskbar.shadow.TaskbarShadowTileService;
import com.farmerbb.taskbar.util.Constants;
import com.farmerbb.taskbar.util.U;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

import static android.content.Context.USER_SERVICE;
import static com.farmerbb.taskbar.Constants.TEST_COMPONENT;
import static com.farmerbb.taskbar.Constants.TEST_LABEL;
import static com.farmerbb.taskbar.Constants.TEST_PACKAGE;
import static com.farmerbb.taskbar.Constants.TEST_WINDOW_SIZE;
import static com.farmerbb.taskbar.util.Constants.EXTRA_COMPONENT_NAME;
import static com.farmerbb.taskbar.util.Constants.EXTRA_PACKAGE_NAME;
import static com.farmerbb.taskbar.util.Constants.EXTRA_USER_ID;
import static com.farmerbb.taskbar.util.Constants.EXTRA_WINDOW_SIZE;
import static com.farmerbb.taskbar.util.Constants.PREF_COMPONENT_NAME_SUFFIX;
import static com.farmerbb.taskbar.util.Constants.PREF_LABEL_SUFFIX;
import static com.farmerbb.taskbar.util.Constants.PREF_PACKAGE_NAME_SUFFIX;
import static com.farmerbb.taskbar.util.Constants.PREF_QS_TILE;
import static com.farmerbb.taskbar.util.Constants.PREF_WINDOW_SIZE_SUFFIX;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;
import static org.robolectric.Shadows.shadowOf;

@RunWith(RobolectricTestRunner.class)
public class FavoriteAppTileServiceTest {
    private FavoriteApp1 app1;
    private SharedPreferences prefs;
    private Context context;

    @Before
    public void setUp() {
        app1 = Robolectric.setupService(FavoriteApp1.class);
        prefs = U.getSharedPreferences(app1);
        context = ApplicationProvider.getApplicationContext();
    }

    @Test
    public void testFavoriteApp1TileNumber() {
        assertEquals(1, app1.tileNumber());
    }

    @Test
    public void testFavoriteApp2TileNumber() {
        FavoriteApp2 app2 = Robolectric.setupService(FavoriteApp2.class);
        assertEquals(2, app2.tileNumber());
    }

    @Test
    public void testFavoriteApp3TileNumber() {
        FavoriteApp3 app3 = Robolectric.setupService(FavoriteApp3.class);
        assertEquals(3, app3.tileNumber());
    }

    @Test
    public void testFavoriteApp4TileNumber() {
        FavoriteApp4 app4 = Robolectric.setupService(FavoriteApp4.class);
        assertEquals(4, app4.tileNumber());
    }

    @Test
    public void testFavoriteApp5TileNumber() {
        FavoriteApp5 app5 = Robolectric.setupService(FavoriteApp5.class);
        assertEquals(5, app5.tileNumber());
    }

    @Test
    public void testOnTileRemoved() {
        app1.onTileRemoved();
        assertFalse(
                prefs.getBoolean(app1.getPrefix() + Constants.PREF_ADDED_SUFFIX, true)
        );
    }

    @Test
    public void testOnClickWithoutAdded() {
        prefs.edit().putBoolean(app1.getPrefix() + Constants.PREF_ADDED_SUFFIX, false).apply();
        app1.onClick();
        Intent startedActivityIntent = shadowOf((Application) context).getNextStartedActivity();
        assertNotNull(startedActivityIntent);
        assertEquals(
                app1.tileNumber(),
                startedActivityIntent.getIntExtra(PREF_QS_TILE, Integer.MIN_VALUE)
        );
        assertNotNull(startedActivityIntent.getComponent());
        assertTrue(
                startedActivityIntent
                        .getComponent()
                        .getClassName()
                        .endsWith(PersistentShortcutSelectAppActivity.class.getSimpleName())
        );
    }

    @Test
    @Config(shadows = {TaskbarShadowTileService.class})
    public void testOnClickWithAdded() {
        prefs.edit().putBoolean(app1.getPrefix() + Constants.PREF_ADDED_SUFFIX, true).apply();

        shadowOf(app1).setLocked(true);
        app1.onClick();
        assertFalse(app1.isLocked());

        shadowOf(app1).setLocked(false);
        String testPackageName = TEST_PACKAGE;
        prefs
                .edit()
                .putString(app1.getPrefix() + PREF_PACKAGE_NAME_SUFFIX, testPackageName)
                .apply();
        String testComponentName = TEST_COMPONENT;
        prefs
                .edit()
                .putString(app1.getPrefix() + PREF_COMPONENT_NAME_SUFFIX, testComponentName)
                .apply();
        String testWindowSize = TEST_WINDOW_SIZE;
        prefs
                .edit()
                .putString(app1.getPrefix() + PREF_WINDOW_SIZE_SUFFIX, testWindowSize)
                .apply();
        app1.onClick();
        Intent startedActivityIntent = shadowOf((Application) context).getNextStartedActivity();
        assertNotNull(startedActivityIntent);
        assertNotNull(startedActivityIntent.getComponent());
        assertTrue(
                startedActivityIntent
                        .getComponent()
                        .getClassName()
                        .endsWith(PersistentShortcutLaunchActivity.class.getSimpleName())
        );
        assertEquals(Intent.ACTION_MAIN, startedActivityIntent.getAction());
        assertEquals(
                testPackageName,
                startedActivityIntent.getStringExtra(EXTRA_PACKAGE_NAME)
        );
        assertEquals(
                testComponentName,
                startedActivityIntent.getStringExtra(EXTRA_COMPONENT_NAME)
        );
        assertEquals(
                testWindowSize,
                startedActivityIntent.getStringExtra(EXTRA_WINDOW_SIZE)
        );
        UserManager userManager = (UserManager) context.getSystemService(USER_SERVICE);
        assertEquals(
                userManager.getSerialNumberForUser(Process.myUserHandle()),
                startedActivityIntent.getLongExtra(EXTRA_USER_ID, Long.MIN_VALUE)
        );
    }

    @Test
    public void testOnStartListeningWithoutAdded() {
        prefs.edit().putBoolean(app1.getPrefix() + Constants.PREF_ADDED_SUFFIX, false).apply();
        app1.onStartListening();
        Tile tile = app1.getQsTile();
        assertEquals(Tile.STATE_INACTIVE, tile.getState());
        assertEquals(app1.getString(R.string.tb_new_shortcut), tile.getLabel());
    }

    @Test
    public void testOnStartListeningWithAdded() {
        prefs.edit().putBoolean(app1.getPrefix() + Constants.PREF_ADDED_SUFFIX, true).apply();
        String testLabel = TEST_LABEL;
        prefs.edit().putString(app1.getPrefix() + PREF_LABEL_SUFFIX, testLabel).apply();
        app1.onStartListening();
        Tile tile = app1.getQsTile();
        assertEquals(Tile.STATE_ACTIVE, tile.getState());
        assertEquals(testLabel, tile.getLabel());
    }
}

File: app/src/test/java/com/farmerbb/taskbar/service/FavoriteAppTileServiceTest.kt
package com.farmerbb.taskbar.service

import android.app.Application
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.os.Process
import android.os.UserManager
import android.service.quicksettings.Tile
import androidx.test.core.app.ApplicationProvider
import com.farmerbb.taskbar.Constants.TEST_COMPONENT
import com.farmerbb.taskbar.Constants.TEST_LABEL
import com.farmerbb.taskbar.Constants.TEST_PACKAGE
import com.farmerbb.taskbar.Constants.TEST_WINDOW_SIZE
import com.farmerbb.taskbar.R
import com.farmerbb.taskbar.activity.PersistentShortcutLaunchActivity
import com.farmerbb.taskbar.activity.PersistentShortcutSelectAppActivity
import com.farmerbb.taskbar.util.Constants
import com.farmerbb.taskbar.util.U
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.Robolectric
import org.robolectric.RobolectricTestRunner
import org.robolectric.Shadows

@RunWith(RobolectricTestRunner::class)
class FavoriteAppTileServiceTest {
    private lateinit var app1: FavoriteApp1
    private lateinit var prefs: SharedPreferences
    private lateinit var context: Context

    @Before
    fun setUp() {
        app1 = Robolectric.setupService(FavoriteApp1::class.java)
        prefs = U.getSharedPreferences(app1)
        context = ApplicationProvider.getApplicationContext()
    }

    @Test
    fun testFavoriteApp1TileNumber() {
        Assert.assertEquals(1, app1.tileNumber().toLong())
    }

    @Test
    fun testFavoriteApp2TileNumber() {
        val app2 = Robolectric.setupService(FavoriteApp2::class.java)
        Assert.assertEquals(2, app2.tileNumber().toLong())
    }

    @Test
    fun testFavoriteApp3TileNumber() {
        val app3 = Robolectric.setupService(FavoriteApp3::class.java)
        Assert.assertEquals(3, app3.tileNumber().toLong())
    }

    @Test
    fun testFavoriteApp4TileNumber() {
        val app4 = Robolectric.setupService(FavoriteApp4::class.java)
        Assert.assertEquals(4, app4.tileNumber().toLong())
    }

    @Test
    fun testFavoriteApp5TileNumber() {
        val app5 = Robolectric.setupService(FavoriteApp5::class.java)
        Assert.assertEquals(5, app5.tileNumber().toLong())
    }

    @Test
    fun testOnTileRemoved() {
        app1.onTileRemoved()
        Assert.assertFalse(
                prefs.getBoolean(app1.prefix + Constants.PREF_ADDED_SUFFIX, true)
        )
    }

    @Test
    fun testOnClickWithoutAdded() {
        prefs.edit().putBoolean(app1.prefix + Constants.PREF_ADDED_SUFFIX, false).apply()
        app1.onClick()
        val startedActivityIntent = Shadows.shadowOf(context as Application?).nextStartedActivity
        Assert.assertNotNull(startedActivityIntent)
        Assert.assertEquals(
                app1.tileNumber().toLong(),
                startedActivityIntent.getIntExtra(Constants.PREF_QS_TILE, Int.MIN_VALUE)
                        .toLong())
        Assert.assertNotNull(startedActivityIntent.component)
        Assert.assertTrue(
                startedActivityIntent
                        .component
                        ?.className
                        ?.endsWith(PersistentShortcutSelectAppActivity::class.java.simpleName)!!
        )
    }

    @Test
    fun testOnClickWithAdded() {
        prefs.edit().putBoolean(app1.prefix + Constants.PREF_ADDED_SUFFIX, true).apply()
        Shadows.shadowOf(app1).setLocked(true)
        app1.onClick()
        Assert.assertFalse(app1.isLocked)
        Shadows.shadowOf(app1).setLocked(false)
        val testPackageName: String = TEST_PACKAGE
        prefs
                .edit()
                .putString(app1.prefix + Constants.PREF_PACKAGE_NAME_SUFFIX, testPackageName)
                .apply()
        val testComponentName: String = TEST_COMPONENT
        prefs
                .edit()
                .putString(app1.prefix + Constants.PREF_COMPONENT_NAME_SUFFIX, testComponentName)
                .apply()
        val testWindowSize: String = TEST_WINDOW_SIZE
        prefs
                .edit()
                .putString(app1.prefix + Constants.PREF_WINDOW_SIZE_SUFFIX, testWindowSize)
                .apply()
        app1.onClick()
        val startedActivityIntent = Shadows.shadowOf(context as Application?).nextStartedActivity
        Assert.assertNotNull(startedActivityIntent)
        Assert.assertNotNull(startedActivityIntent.component)
        Assert.assertTrue(
                startedActivityIntent
                        .component
                        ?.className
                        ?.endsWith(PersistentShortcutLaunchActivity::class.java.simpleName)!!
        )
        Assert.assertEquals(Intent.ACTION_MAIN, startedActivityIntent.action)
        Assert.assertEquals(
                testPackageName,
                startedActivityIntent.getStringExtra(Constants.EXTRA_PACKAGE_NAME)
        )
        Assert.assertEquals(
                testComponentName,
                startedActivityIntent.getStringExtra(Constants.EXTRA_COMPONENT_NAME)
        )
        Assert.assertEquals(
                testWindowSize,
                startedActivityIntent.getStringExtra(Constants.EXTRA_WINDOW_SIZE)
        )
        val userManager = context.getSystemService(Context.USER_SERVICE) as UserManager
        Assert.assertEquals(
                userManager.getSerialNumberForUser(Process.myUserHandle()),
                startedActivityIntent.getLongExtra(Constants.EXTRA_USER_ID, Long.MIN_VALUE)
        )
    }

    @Test
    fun testOnStartListeningWithoutAdded() {
        prefs.edit().putBoolean(app1.prefix + Constants.PREF_ADDED_SUFFIX, false).apply()
        app1.onStartListening()
        val tile = app1.qsTile
        Assert.assertEquals(Tile.STATE_INACTIVE.toLong(), tile.state.toLong())
        Assert.assertEquals(app1.getString(R.string.tb_new_shortcut), tile.label)
    }

    @Test
    fun testOnStartListeningWithAdded() {
        prefs.edit().putBoolean(app1.prefix + Constants.PREF_ADDED_SUFFIX, true).apply()
        val testLabel: String = TEST_LABEL
        prefs.edit().putString(app1.prefix + Constants.PREF_LABEL_SUFFIX, testLabel).apply()
        app1.onStartListening()
        val tile = app1.qsTile
        Assert.assertEquals(Tile.STATE_ACTIVE.toLong(), tile.state.toLong())
        Assert.assertEquals(testLabel, tile.label)
    }
}
